_posts/2019/2019-02-10-be-careful-assigning-to-has-one-relations.md

```diff
@@ -0,0 +1,132 @@
+=> #<Biography id: 2, person_id: 1, text: Looks handsome ...>
+Biography.count
+=> 1
+```
+
+Boom.
+
+
+## But why?
+
+The first example demonstrates that the assignment (in this case via `#biography=`) to the `has_one` attribute removes the previously associated model and writes to the database.
+
+In Rails prior to 5.2.1 there is a bug that _even if_ the newly assigned model was invalid and would not save the existing record would still be changed.
+
+Normally when making assignments to the attributes of an Active Record model, via a `#whatever=` method, the changes are not persisted to the database until you call `#save`. In this association case, however, the change to the previously assigned model is written _at the point of assignment_.
+
+The method I’ve created in the second solution uses the `!` variants of the `create` and `destroy` methods inside a transaction to raise exceptions and rollback both the unassignment of the existing model and assignment of the new one if either fails.
+
+This prevents subtle bugs and the unusual behaviour of implicit writing to the database without calling `#save`.
+
+
+### In Modern Rails (after 5.0)
+
+In Rails 5.0, if you attempt to set the `person_id` attribute to `nil`, as the assignation of the new model will do, an `ActiveRecord::RecordNotSaved` error is raised. This prevents the assignation even if the new record is valid.
+
+The model being removed from the association is no longer valid because `belongs_to` relationships now have an [implied presence validation](https://guides.rubyonrails.org/5_0_release_notes.html#active-record-notable-changes). The change in Rails’ behaviour happened in [this pull request](https://github.com/rails/rails/pull/18937).
+
+I used the `optional: true` argument, in my example, to make the `Person` class behave more like pre-5.0 Rails.
+
+The solution I’ve proposed would also work in this more strict `belongs_to` relationship.
+
+
+## Why not?
+
+You can manage the `has_one` relationship in multiple other ways that do not wrap the whole task in a `replace_whatever`-style method such as I’ve suggested.
+
+In fact this [bug fix in Rails 5.2.1](https://github.com/rails/rails/pull/32796) ([commit](https://github.com/rails/rails/commit/c87b3346ca6e1d21a6bccb29ccedf0b95fda7abc)) means the `#whatever=` method does now rollback the changes as my solution would.
+
+You could:
+
+  * Avoid calling the `#whatever=` method in your code and use the clearer `#create_whatever!()` instead
+  * Allow assignment only when the association is empty
+  * Declare your association with `autosave: true/false` to make the behaviour explicit in your code
+
+...but you will need to manage it somehow.
+
+There is a reasonable argument that using `has_one` leads to more edge cases than the more typical `has_many` relationship. One approach would be to engineer the logic of your application to use that relationship, which might make for more predictable behaviour.
```

Comment:
I have to say — I get to the end of this article and I'm like: wait — so what am I meant to do? Which Rails version am I on? What is best?

Admittedly, I started reading it last night and finished it this afternoon...so I've had a context gap...but there is a lot here.

Reply:
There is. But it's all interrelated.

Reply:
https://github.com/andycroll/andycroll.com/pull/38/commits/c94fb096d0936f2542bc86203bc8af667ad17a20 < worth it I think

Reply:
Instead, maybe it's after the list, where you say: "...but you have to deal with it somehow." Maybe you point out there: the main thing I want you to take away is: be aware of the peculiarities (don't say it exactly like that, but you catch my drift!)

Reply:
Done.